README
Playground for Jadex multi-agent platform and simulator.


jadex/changes.txt
jadex change log.
$Revision: 5271 $
$Date: 2007-06-13 22:08:29 +0200 (Wed, 13 Jun 2007) $

Changes for release "Jadex-0.96" (2007/06/15):

06/12/2007
	- Checkbox for saving platform shutdown option on exit.
	- Saving/restoring console divider location individually for each plugin.
	- Jadexdoc plugin shows Java sources for generating pure javadoc.

05/29/2007
	- Fixed some problems related to using jar files in starter.

05/25/2007
	- Fixed some problems related to closing the platform when tools are open.


Changes for beta release "Jadex-0.96-beta1" (2006/03/20):

03/20/2007
	- Fixed protocols capability for auctions.

03/19/2007
	- Fixed bug in process goal parameter sets.

03/12/2007
	- Beliefbase panel introspector provides checking feedback while typing belief values.

03/11/2007
	- Changed goalbase view in introspector (introduced goal/plan view).
	- Introspector now visualizes which agents are introspected.

03/06/2007
	- New Jadexdoc plugin introduced. Allows easy drag and drop generation of Jadexdocs.
	- Added a console output to the JCC. 

03/02/2007
	- Speedup of JCC loading due to lazy initialization of plugins.

02/22/2007
	- Refactored examples to take advantage of new agent-level listener features.

02/14/2007
	- Introduced end state containing goals, plans and events, which are created
		on agent termination.

02/12/2007
	- Agent arguments can be supplied as name-value (expression) pairs, arguments are bound
		to exported agent beliefs, starter shows and allows to enter arguments one by one 

02/11/2007
	- Now storing duplicity settings in testcenter.
	- Added add/remove testcase buttons in test center toolbar (replacing arrow besides list).

02/08/2007
	- Made $addedfact and $removedfact accessible in ADFs of plans (cf. user request sourceforge)
	- provided access to agent arguments from plan / guiwrapper via getArguments() -> feature request
	- Now checking for correct types of expressions.

02/07/2007
	- Introduced callback listeners for agent actions
	- Cleaned-up exteranl access and plan interfaces
	- Now checking for correct beliefs/beliefsets in initial states
	- Now checking correct name of file (disabled auto correction).

02/05/2007
	- Now allowing execution of plans and goals during agent termination.

02/02/2007
	- Now allowing subgoals in passed/failed/aborted.

02/01/2007
	- Fixed bug in XML model loading, preventing input stream from being closed.

01/26/2007
	- Fixed bug in package check regarding jar files in classpath.
	- Fixed bug in test suite load/save dialog (no directories displayed).
	- Initial elements can have name (and later be referenced by)
	- cleaned up binding issues (calculateCartesianProduct() in SUtil)

01/05/2007
	- Improved starter popups for starting.
	- Improved test center, forbid duplicates, save report, add directories.
	- Passed/failed/aborted are now treated them as normal plan steps (i.e. support waitFor).

01/02/2007
	- Wrong packages are now detected in error reports.
	- Dynamic pop-ups in starter/test-center trees.
	- Pop-up in starter to select initial configuration.
	
12/30/2006
	- implement message forwarding (Morandini)

12/29/2006
	- Implemented repeatability for condition execution.

12/23/2006
	- Introduced one click agent starting (double click on model name) by adding
		a auto generate agent name flag in starter. 

12/22/2006
	- Saving of model tree now also works for jar files.

12/19/2006
	- Fixed bug in timetable related to long waiting times (e.g. in waitFor() calls).
	- Finished initial version of iterated contract net protocol

12/17/2006
	- Starter/testcenter now save checking state in model tree for improved startup speed.

12/15/2006
	- Fixed swing worker for starter panel.
	- Changed contract net allowing to have several winners.
	- Model cache now uses LRU scheme.

12/14/2006
	- Fixed loading/saving/resetting of settings in testcenter.
	- Fixed loading/saving/resetting of settings in conv center.
	- Open/save project in JCC defaults to current directory.


Changes for beta release "Jadex-0.95-beta1" (2006/07/07):

07/06/2006
	- Fixed bug in model explorer used by starter panel and test center.

06/23/2006
	- Added strong exports as default and introduced new keyword ("shielded")
	    for old weak exports.
	- Initial version of protocols capability currently supporting request and contract-net protocols.
	- New booktrading example as described in JADE book chapter on Jadex.

06/13/2006
	- Fixed bug in urlclassloader causing deadlocks.

06/12/2006
	- Fixed bug in (javacc) parser concerning nonpublic classes.
	- Added custom match expressions for message events.
	- Added custom match expressions for plan trigger (e.g. goals or events).

06/09/2006
	- Introduced recur, recurdelay for all goals (not only maintain).

06/06/2006
	- Fixed bug regarding message event references.

04/28/2006
	- Added some more checks to model (e.g. ref and class attributes)
	- Fixed CPU consumption of background checking of models

03/30/2006
	- Now top-level goal/plan failures produce warning messages.

03/29/2006
	- Added DFBrowser plugin to inspect registered agents.

03/28/2006
	- Fixed bugs in conversation center and agent selector dialog.

03/24/2006
	- BrowserPane now also searches the classpath for images.
	- Fixed bug in maintain goal retry.

03/22/2006
	- Fixed minor issue concerning info events of subgoals of finished plans.

03/21/2006
	- Added isEmpty/size methods to waitqueue.
	- Fixed bug in abstract goal parameters.

03/20/2006
	- Added RMAPlugin to show JADE RMA in JCC.
	- Added first version of test suite support.

02/16/2006
	- Implemented nugget codec as replacement for the slow java xml codec.
	- Added indication in JCC when scanning for models is in progress.

02/14/2006
	- Introduced support for custom content codecs for handling (fipa) message content.

01/20/2006
	- Removed unnecessary notifyDue() when last timing entry was removed.

01/18/2006

jadex/src/jadex/util/Base64.java
package jadex.util;

/** Base64<br>
 * Contains lots of utilities for coding and decoding text and character arrays fore
 * and back to the Base64 encoding. It is quite the fastest that I know.
 * @author Walczak
 * @since Sep 8, 2005
 * @version 1.1
 */
public final class Base64
{
   // ------------------------ ENCODER --------------------
   /** This is a wrapper to <code>toCharArray(final byte[] text, final int from, final int len)</code>
    * @param text
    * @return an encoded char array from the text
    */
   public final static char[] toCharArray(final byte[] text)
   {
      return toCharArray(text, 0, text.length);
   }

   /** This will encode the text without line feeds added
    * @param text
    * @param from where to start
    * @param len how long is the byte array
    * @return an encoded char array from the text
    */
   public final static char[] toCharArray(final byte[] text, final int from, final int len)
   {
      final char[] code;
      int bi = from + len;
      int ci = (len / 3) << 2;
      int i;
      switch (len % 3) { // length of padding
      case 1:
         ci += 4;
         code = new char[ci];
         i = text[--bi] << 4;
         code[--ci] = '=';
         code[--ci] = '=';
         code[--ci] = C[i & 0x3F];
         code[--ci] = C[(i >>> 6) & 0x3F];
         break;
      case 2:
         ci += 4;
         code = new char[ci];
         i = (text[--bi] & 0xFF) << 2 | text[--bi] << 10;
         code[--ci] = '=';
         code[--ci] = C[i & 0x3F];
         code[--ci] = C[(i >>> 6) & 0x3F];
         code[--ci] = C[(i >>> 12) & 0x3F];
         break;
      default: // 0
         code = new char[ci];
      }

      while (ci > 00)
      {
         i = (text[--bi] & 0xFF) | (text[--bi] & 0xFF) << 8 | text[--bi] << 16;
         code[--ci] = C[i & 0x3F];
         code[--ci] = C[(i >>> 6) & 0x3F];
         code[--ci] = C[(i >>> 12) & 0x3F];
         code[--ci] = C[(i >>> 18) & 0x3F];
      }
      return code;
   }

   /** 
    * @param text
    * @return encoded text
    */
   public final static byte[] encode(final byte[] text)
   {
      return encode(text, 0, text.length);
   }

   /** This will encode the text without line feeds added
    * @param text
    * @param from where to start
    * @param len how long is the byte array
    * @return an encoded byte array from the text
    */
   public final static byte[] encode(final byte[] text, final int from, final int len)
   {
      final byte[] code;
      int bi = from + len;
      int ci = (len / 3) << 2;
      int i;
      switch (len % 3) { // length of padding
      case 1:
         ci += 4;
         code = new byte[ci];
         i = text[--bi] << 4;
         code[--ci] = '=';
         code[--ci] = '=';
         code[--ci] = B[i & 0x3F];
         code[--ci] = B[(i >>> 6) & 0x3F];
         break;
      case 2:
         ci += 4;
         code = new byte[ci];
         i = (text[--bi] & 0xFF) << 2 | text[--bi] << 10;
         code[--ci] = '=';
         code[--ci] = B[i & 0x3F];
         code[--ci] = B[(i >>> 6) & 0x3F];
         code[--ci] = B[(i >>> 12) & 0x3F];
         break;
      default: // 0
         code = new byte[ci];
      }

      while (ci > 00)
      {
         i = (text[--bi] & 0xFF) | (text[--bi] & 0xFF) << 8 | text[--bi] << 16;
         code[--ci] = B[i & 0x3F];
         code[--ci] = B[(i >>> 6) & 0x3F];
         code[--ci] = B[(i >>> 12) & 0x3F];
         code[--ci] = B[(i >>> 18) & 0x3F];
      }
      return code;
   }

   /** This will encode the text without line feeds added. It will write all output
    * to the dest buffer begining from dstart.<br>
    * Both src and dest may be the same provided (sstart<=dstart).
    * @param src
    * @param sstart where to start
    * @param len how long is the byte array
    * @param dest
    * @param dstart
    * @return index to the buffer byte ofter the text
    */

   public final static int encode(final byte[] src, final int sstart, final int len, final byte[] dest, final int dstart)
   {
      int bi = sstart + len;
      int ci = ((len / 3) << 2) + dstart;
      int dend;
      int i;
      switch (len % 3) { // length of padding
      case 1:
         ci += 4;
         dend = ci;
         i = src[--bi] << 4;
         dest[--ci] = '=';
         dest[--ci] = '=';
         dest[--ci] = B[i & 0x3F];
         dest[--ci] = B[(i >>> 6) & 0x3F];
         break;
      case 2:
         ci += 4;
         dend = ci;
         i = (src[--bi] & 0xFF) << 2 | src[--bi] << 10;
         dest[--ci] = '=';
         dest[--ci] = B[i & 0x3F];
         dest[--ci] = B[(i >>> 6) & 0x3F];
         dest[--ci] = B[(i >>> 12) & 0x3F];
         break;
      default:
         dend = ci;
      }

      while (ci > dstart)
      {
         i = (src[--bi] & 0xFF) | (src[--bi] & 0xFF) << 8 | src[--bi] << 16;
         dest[--ci] = B[i & 0x3F];
         dest[--ci] = B[(i >>> 6) & 0x3F];
         dest[--ci] = B[(i >>> 12) & 0x3F];
         dest[--ci] = B[(i >>> 18) & 0x3F];
      }
      return dend;
   }

   /** This will encode the text. The CRLF comes after <code>lb</code> characters.
    * @param text
    * @param llen the length of line (without CRLF) must be a multiple of 4
    * @return an encoded char array from the text
    */
   public final static char[] toCharArray(final byte[] text, final int llen)
   {
      final char[] code;
      int bi = text.length;
      int ci = (bi / 3) << 2;
      int i;
      //@goal determine padding
      switch (bi % 3) { // length of padding
      case 1: // @goal pad two text
         ci += 4;
         ci += (ci / llen) << 1;
         code = new char[ci];

         if (ci % llen == 4)
         {
            code[--ci] = '\n';
            code[--ci] = '\r';
         }
         i = text[--bi] << 4;
         code[--ci] = '=';
         code[--ci] = '=';
         code[--ci] = C[i & 0x3F];
         code[--ci] = C[(i >>> 6) & 0x3F];
         break;
      case 2: // @goal pad one byte
         ci += 4;
         ci += (ci / llen) << 1;
         code = new char[ci];

         if (ci % llen == 4)
         {
            code[--ci] = '\n';
            code[--ci] = '\r';
         }
         i = (text[--bi] & 0xFF) << 2 | text[--bi] << 10;
         code[--ci] = '=';
         code[--ci] = C[i & 0x3F];
         code[--ci] = C[(i >>> 6) & 0x3F];
         code[--ci] = C[(i >>> 12) & 0x3F];
         break;
      default: // 0
         ci += (ci / llen) << 1;
         code = new char[ci];
      }
      //@goal calculate the line breaking constant and variable
      int lmax = (llen >>> 2) - 1; // will break after this many quadruples
      int li = (ci % (llen + 2)) >>> 2; // initialize the counter

      while (bi > 00)
      {

         if (li > 0)
         {
            --li;

         }
         else
         {
            code[--ci] = '\n';
            code[--ci] = '\r';
            li = lmax;
         }
         i = (text[--bi] & 0xFF) | (text[--bi] & 0xFF) << 8 | text[--bi] << 16;
         code[--ci] = C[i & 0x3F];
         code[--ci] = C[(i >>> 6) & 0x3F];
         code[--ci] = C[(i >>> 12) & 0x3F];
         code[--ci] = C[(i >>> 18) & 0x3F];
      }
      return code;
   }

   /** This will encode the text. The CRLF comes after <code>lb</code> characters.
    * @param text
    * @param llen the length of line (without CRLF) must be a multiple of 4
    * @return an encoded byte array from the text
    */

   public final static byte[] encode(final byte[] text, final int llen)
   {
      final byte[] code;
      int bi = text.length;
      int ci = (bi / 3) << 2;
      int i;
      //@goal determine padding
      switch (bi % 3) { // length of padding
      case 1: // @goal pad two text
         ci += 4;
         ci += (ci / llen) << 1;
         code = new byte[ci];

         if (ci % llen == 4)
         {
            code[--ci] = '\n';
            code[--ci] = '\r';
         }
         i = text[--bi] << 4;
         code[--ci] = '=';
         code[--ci] = '=';
         code[--ci] = B[i & 0x3F];
         code[--ci] = B[(i >>> 6) & 0x3F];
         break;
      case 2: // @goal pad one byte
         ci += 4;
         ci += (ci / llen) << 1;
         code = new byte[ci];

         if (ci % llen == 4)
         {
            code[--ci] = '\n';
            code[--ci] = '\r';
         }
         i = (text[--bi] & 0xFF) << 2 | text[--bi] << 10;
         code[--ci] = '=';
         code[--ci] = B[i & 0x3F];
         code[--ci] = B[(i >>> 6) & 0x3F];
         code[--ci] = B[(i >>> 12) & 0x3F];
         break;
      default: // 0
         ci += (ci / llen) << 1;
         code = new byte[ci];
      }
      //@goal calculate the line breaking constant and variable
      int lmax = (llen >>> 2) - 1; // will break after this many quadruples
      int li = (ci % (llen + 2)) >>> 2; // initialize the counter

      while (ci > 00)
      {

         if (li > 0)
         {
            --li;

         }
         else
         {
            code[--ci] = '\n';
            code[--ci] = '\r';
            li = lmax;
         }
         i = (text[--bi] & 0xFF) | (text[--bi] & 0xFF) << 8 | text[--bi] << 16;
         code[--ci] = B[i & 0x3F];
         code[--ci] = B[(i >>> 6) & 0x3F];
         code[--ci] = B[(i >>> 12) & 0x3F];
         code[--ci] = B[(i >>> 18) & 0x3F];
      }
      return code;
   }

   //###################### DECODER #####################################
   /** This is a wrapper to <code>decode(final char[] code, final int from, final int len)</code>
    * @param code
    * @return a byte array holding encoded char data
    */
   public final static byte[] decode(final char[] code)
   {
      return decode(code, 0, code.length);
   }

   /** This will decode base64 data without line feeds.
    * The char array should be multiple of 4 in length.
    * @param code
    * @param from start position in code
    * @param len length of the part
    * @return the decoded sequence of text
    */
   public final static byte[] decode(final char[] code, final int from, final int len)
   {
      int bi = (len >>> 2) * 3;
      int ci = from + len;
      int i;
      final byte[] text;

      if (code[ci - 1] == '=')
      {
         if (code[ci - 2] == '=')
         { // padding 2
            ci -= 3;
            bi -= 2;
            text = new byte[bi];
            text[--bi] = (byte) ((I[code[ci]] | I[code[--ci]] << 6) >> 4);
         }
         else
         { // padding 1
            ci -= 2;
            --bi;
            text = new byte[bi];
            i = I[code[ci]] | I[code[--ci]] << 6 | I[code[--ci]] << 12;
            text[--bi] = (byte) (i >>> 2);
            text[--bi] = (byte) (i >>> 10);
         }

      }
      else
      {
         text = new byte[bi];
      }

      while (bi > 00)
      {
         i = I[code[--ci]] | I[code[--ci]] << 6 | I[code[--ci]] << 12 | I[code[--ci]] << 18;
         text[--bi] = (byte) (i);
         text[--bi] = (byte) (i >>> 8);
         text[--bi] = (byte) (i >>> 16);
      }
      return text;
   }

   /** This is a wrapper to <code>decode(final byte[] code, final int from, final int len)</code>
    * @param code
    * @return a byte array holding encoded char data
    */
   public final static byte[] decode(final byte[] code)
   {
      return decode(code, 0, code.length);
   }

   /** a copy of the char part
    * @param src
    * @param from
    * @param len
    * @return decoded byte array
    */
   public final static byte[] decode(final byte[] src, final int from, final int len)
   {
      int bi = (len >>> 2) * 3;
      int ci = from + len;
      int i;
      final byte[] text;

      if (src[ci - 1] == '=')
      {
         if (src[ci - 2] == '=')
         { // padding 2
            ci -= 3;
            bi -= 2;
            text = new byte[bi];
            text[--bi] = (byte) ((I[src[ci]] | I[src[--ci]] << 6) >> 4);
         }
         else
         { // padding 1
            ci -= 2;
            --bi;
            text = new byte[bi];
            i = I[src[ci]] | I[src[--ci]] << 6 | I[src[--ci]] << 12;
            text[--bi] = (byte) (i >>> 2);
            text[--bi] = (byte) (i >>> 10);
         }

      }
      else
      {
         text = new byte[bi];
      }

      while (bi > 00)
      {
         i = I[src[--ci]] | I[src[--ci]] << 6 | I[src[--ci]] << 12 | I[src[--ci]] << 18;
         text[--bi] = (byte) (i);
         text[--bi] = (byte) (i >>> 8);
         text[--bi] = (byte) (i >>> 16);
      }
      return text;
   }

   /** <code>decodeCRLF(code, 0, code.length, 76)</code>
    * @param code
    * @return the encoded text;
    */
   public final static byte[] decode76(final char[] code)
   {
      return decodeCRLF(code, 0, code.length, 76);
   }

   /** This will decode base64 data. The starting point and length must be accurate.
    * The data must end on a multiple of 4 boundary and must include the '=' padding, if any.
    * @param code
    * @param from
    * @param len the length of data
    * @param llen the line length of this base64 (without CRLF)
    * @return the decoded sequence of text
    */
   public final static byte[] decodeCRLF(final char[] code, final int from, final int len, final int llen)
   {
      int bi = ((len - ((len / (llen + 2)) << 1)) >>> 2) * 3;
      int ci = from + len;
      int lmax = (llen >>> 2);
      int li = (ci % (llen + 2)) >>> 2;
      int i;
      if (li == 0)
      { // skip crlf
         ci -= 2;
         li = lmax;
      }
      final byte[] text;

      if (code[ci - 1] == '=')
      {
         if (code[ci - 2] == '=')
         { // padding 2
            ci -= 3;
            bi -= 2;
            text = new byte[bi];
            text[--bi] = (byte) ((I[code[ci]] | I[code[--ci]] << 6) >> 4);
            --li;
         }
         else
         { // padding 1
            ci -= 2;
            --bi;
            text = new byte[bi];
            i = I[code[ci]] | I[code[--ci]] << 6 | I[code[--ci]] << 12;
            text[--bi] = (byte) (i >>> 2);
            text[--bi] = (byte) (i >>> 10);
            --li;
         }

      }
      else
      {
         text = new byte[bi];
      }

      while (bi > 00)
      {

         if (li == 0)
         {
            ci -= 2;
            li = lmax;
         }
         i = I[code[--ci]] | I[code[--ci]] << 6 | I[code[--ci]] << 12 | I[code[--ci]] << 18;
         text[--bi] = (byte) (i);
         text[--bi] = (byte) (i >>> 8);
         text[--bi] = (byte) (i >>> 16);
         --li;
      }
      return text;
   }

   /** This will decode base64 data with CRLF at 4 character boundary.<br>
    * The sequence may look like:<p>
    * ABCDABCD\r\nABCDABCD or even ABCDABCD#####ABCD###ABCD#ABCD.</p>
    * <p>
    * The array must be multiple of 4 + number of CRLF or illegal characters and 
    * the line length may vary.</p>
    * @param code
    * @param from
    * @param len
    * @return the decoded sequence of text
    */
   public final static byte[] decodeCRLF(final char[] code, final int from, final int len)
   {
      int bi = len;
      int ci = from + len;
      int i;
      //@goal determine the number of valid code

      while (ci > from)
      {
         if (I[code[--ci]] < 0) --bi;
      }
      //@goal allocate byte array
      bi = (bi >>> 2) * 3;
      final byte[] text = new byte[bi];
      //@goal decode the sequence
      ci = from + len;

      while (bi > 00)
      {
         do
            i = I[code[--ci]]; // look ahead
         while (i < 0);
         i |= I[code[--ci]] << 6 | I[code[--ci]] << 12 | I[code[--ci]] << 18;
         text[--bi] = (byte) (i);
         text[--bi] = (byte) (i >>> 8);
         text[--bi] = (byte) (i >>> 16);
      }
      return text;
   }

   /** 
    * @param code
    * @return the decoded array
    */
   public final static byte[] decodeFailSafe(char[] code)
   {
      return decodeFailSafe(code, 0, code.length);
   }

   /** This removes all bad characters from the char array. It modifies the array
    * in the scope of the process !!! Than simply calls decode;
    * @param code
    * @param from
    * @param len
    * @return decoded text
    */
   public final static byte[] decodeFailSafe(final char[] code, final int from, final int len)
   {
      char c;
      int ci = from;
      int cj = from;
      int ce = from + len;

      while (ci < ce)
      {
         c = code[ci];
         if (c == '=') break;

         try
         {
            if (I[c] >= 0) code[cj++] = c;
         }
         catch (Exception e)
         {/**/}
         ci++;
      }

      switch (((cj - from) & 0x3)) {
      case 1:
         code[cj++] = C[0];
      case 2:
         code[cj++] = '=';
      case 3:
         code[cj++] = '=';
      default:
      }
      return decode(code, from, cj - from);
   }

   //------------------ static fields -------------------

   private static final byte[] B = new byte[] { 
         65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 
         81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 97, 98, 99,100,101,102,
        103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,
        119,120,121,122, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 43, 47, };

   private static final char[] C = new char[] { 
        'A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P',
        'Q','R','S','T','U','V','W','X','Y','Z','a','b','c','d','e','f',
        'g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v',
        'w','x','y','z','0','1','2','3','4','5','6','7','8','9','+','/', };

   private static final int[]  I = new int[] { 
         -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
         -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
         -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 62, -1, -1, -1, 63,
         52, 53, 54, 55, 56, 57, 58, 59, 60, 61, -1, -1, -1,  0, -1, -1,
         -1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 
         15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, -1, -1, -1, -1, -1, 
         -1, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 
         41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, -1, -1, -1, -1, -1, };
}